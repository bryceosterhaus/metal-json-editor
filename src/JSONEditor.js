'use strict';

import Component, {Config} from 'metal-jsx';

import Tree from './Tree';
import {setIn} from './util';

class JSONEditor extends Component {
  handleChange(locator, value) {
    const {data, onChange} = this.props;

    onChange(setIn(data, locator, value));
  }

  handleToggleExpand(value) {
    const {expandedNodes} = this.state;

    if (expandedNodes.has(value)) {
      expandedNodes.delete(value);
    } else {
      expandedNodes.add(value);
    }

    this.state.expandedNodes = expandedNodes;
  }

  render() {
    const {arrowRenderer, data, typeColors} = this.props;

    return (
      <Tree
        arrowRenderer={arrowRenderer}
        data={data}
        expandedNodes={this.state.expandedNodes}
        onChange={this.handleChange.bind(this)}
        onToggleExpand={this.handleToggleExpand.bind(this)}
        typeColors={typeColors}
      />
    );
  }
}

JSONEditor.STATE = {
  expandedNodes: Config.instanceOf(WeakSet).value(new WeakSet())
};

JSONEditor.PROPS = {
  arrowRenderer: Config.func().value(() => {}),
  data: Config.object().value({}),
  onChange: Config.func(),
  typeColors: Config.object()
};

export default JSONEditor;
