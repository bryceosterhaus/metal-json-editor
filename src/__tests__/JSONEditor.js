import JSONEditor from '../JSONEditor';

describe('JSONEditor', () => {
  let component;

  afterEach(() => {
    if (component) {
      component.dispose();
    }
  });

  it('renders', () => {
    component = new JSONEditor();

    expect(component).toMatchSnapshot();
  });

  it('renders with custom typeColors', () => {
    component = new JSONEditor({
      data: {
        bar: 1,
        baz: true,
        foo: 'bar'
      }
    });

    expect(component).toMatchSnapshot();

    const componentWithColors = new JSONEditor({
      data: {
        bar: 1,
        baz: true,
        foo: 'bar'
      },
      typeColors: {
        boolean: '#000',
        number: '#111',
        string: '#222'
      }
    });

    expect(componentWithColors).toMatchSnapshot();
  });

  it('should call onChange', () => {
    const spy = jest.fn();

    component = new JSONEditor({
      data: {
        foo: 'bar'
      },
      onChange: spy
    });

    component.handleChange(['foo'], 'baz');

    expect(spy.mock.calls).toMatchSnapshot();
  });

  it('should set and remove expanded value from "expandedNodes"', () => {
    component = new JSONEditor();

    component.state.expandedNodes = new WeakSet();

    const obj = {foo: 'foo'};

    component.handleToggleExpand(obj);

    jest.runAllTimers();

    expect(component.state.expandedNodes.has(obj)).toMatchSnapshot();

    component.handleToggleExpand(obj);

    jest.runAllTimers();

    expect(component.state.expandedNodes.has(obj)).toMatchSnapshot();
  });
});
