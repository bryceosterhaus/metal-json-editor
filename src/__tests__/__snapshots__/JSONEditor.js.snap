// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONEditor renders 1`] = `<ul class="tree-container" style="list-style: none; margin: 0px;"></ul>`;

exports[`JSONEditor renders with custom typeColors 1`] = `
<ul class="tree-container" style="list-style: none; margin: 0px;">
  <li key="_metal_jsx_1">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">bar: </span>
      <span style="color: rgb(28, 0, 207);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">1</span>
      </span>
    </div>
  </li>
  <li key="_metal_jsx_2">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">baz: </span>
      <span style="color: rgb(170, 13, 145);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">true</span>
      </span>
    </div>
  </li>
  <li key="_metal_jsx_3">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">foo: </span>
      <span style="color: rgb(196, 26, 22);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">bar</span>
      </span>
    </div>
  </li>
</ul>
`;

exports[`JSONEditor renders with custom typeColors 2`] = `
<ul class="tree-container" style="list-style: none; margin: 0px;">
  <li key="_metal_jsx_1">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">bar: </span>
      <span style="color: rgb(17, 17, 17);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">1</span>
      </span>
    </div>
  </li>
  <li key="_metal_jsx_2">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">baz: </span>
      <span style="color: rgb(0, 0, 0);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">true</span>
      </span>
    </div>
  </li>
  <li key="_metal_jsx_3">
    <div key="_metal_jsx_1">
      <span style="" key="_metal_jsx_1">foo: </span>
      <span style="color: rgb(34, 34, 34);" title="Click to Edit" key="_metal_jsx_2">
        <span style="white-space: pre;" key="_metal_jsx_1">bar</span>
      </span>
    </div>
  </li>
</ul>
`;

exports[`JSONEditor should call onChange 1`] = `
Array [
  Array [
    Object {
      "foo": "baz",
    },
  ],
]
`;

exports[`JSONEditor should set and remove expanded value from "expandedNodes" 1`] = `true`;

exports[`JSONEditor should set and remove expanded value from "expandedNodes" 2`] = `false`;
